Assets/__Scripts/UI/SongSelectMenu/SongList.cs

```diff
@@ -1,77 +1,137 @@
+    private List<BeatSaberSong> _filteredSongs = new List<BeatSaberSong>();
+    [SerializeField] private RecyclingListView newList;
+
+    private void SwitchSort(IComparer<BeatSaberSong> newSort, Sprite sprite)
+    {
+        sortImage.sprite = sprite;
+        Songs = new SortedSet<BeatSaberSong>(Songs, newSort);
+        UpdateSongList();
+    }
 
+    public void NextSort()
+    {
+        switch (_currentSort)
+        {
+            case SongSortType.Name:
+                SwitchSort(SortModified, modifiedSortSprite);
+                _currentSort = SongSortType.Modified;
+                break;
+            case SongSortType.Modified:
+                SwitchSort(SortArtist, artistSortSprite);
+                _currentSort = SongSortType.Artist;
+                break;
+            default:
+                SwitchSort(SortName, nameSortSprite);
+                _currentSort = SongSortType.Name;
+                break;
+        }
+    }
+    
     private void Start()
     {
-        WIPLevels = lastVisited_WasWIP;
-        RefreshSongList();
+        newList.ItemCallback = (item, index) =>
+        {
+            if (item is SongListItem child)
+            {
+                child.AssignSong(_filteredSongs[index], searchField.text);
+            }
+        };
+        
+        WipLevels = _lastVisitedWasWip;
+        TriggerRefresh();
     }
 
     public void ToggleSongLocation()
     {
-        WIPLevels = !WIPLevels;
-        lastVisited_WasWIP = WIPLevels;
-        RefreshSongList();
+        SetSongLocation(!WipLevels);
     }
 
-    public void RefreshSongList() {
-        songLocationToggleText.StringReference.TableEntryReference = WIPLevels ? "custom" : "wip";
+    public void SetSongLocation(bool wip)
+    {
+        _lastVisitedWasWip = WipLevels = wip;
+        wipTab.color = wip ? selectedTabColor : normalTabColor;
+        customTab.color = !wip ? selectedTabColor : normalTabColor;
+        TriggerRefresh();
+    }
 
-        FilteredBySearch = !string.IsNullOrEmpty(searchField.text);
-        string[] directories;
-        directories = Directory.GetDirectories(WIPLevels ? Settings.Instance.CustomWIPSongsFolder : Settings.Instance.CustomSongsFolder);
-        songs.Clear();
+    public void TriggerRefresh()
+    {
+        StartCoroutine(RefreshSongList());
+    }
+
+    public IEnumerator RefreshSongList() {
+        var directories = Directory.GetDirectories(WipLevels ? Settings.Instance.CustomWIPSongsFolder : Settings.Instance.CustomSongsFolder);
+        Songs.Clear();
+        newList.Clear();
+        var iterBeginTime = Time.realtimeSinceStartup;
         foreach (var dir in directories)
         {
-            BeatSaberSong song = BeatSaberSong.GetSongFromFolder(dir);
+            if (Time.realtimeSinceStartup - iterBeginTime > 0.02f)
+            {
+                UpdateSongList();
+                yield return null;
+                iterBeginTime = Time.realtimeSinceStartup;
+            }
```

Comment:
ooc, what's the purpose of this bit?

Reply:
To prevent stalling the main thread, I have >400 maps in custom levels and that takes more than 1 frame to load.
But if we yield every loop then it takes >400 frames to load, this tries to find a middle ground.

I guess the alternative is to put it all on another thread and then poll every frame until it completes... 🤔

Reply:
nothing needs to be done, i was just wondering the purpose behind it